회원가입 화면 #step3 jake

jryoun1 opened the pull request:
step3 구현했습니다.
 
[고민한 부분]
1. 처음에는 step3 화면에서 사용자로부터 입력을 받는 즉시 UserInformation의 shared에 저장을 하려고했는데, 이 부분에 대해서 고민을 해본 결과 회원가입 버튼을 눌렀을 때 저장을 하는 것이 바람직한 것 같다고 생각하였습니다. 따라서 step3, step4 에서의 사용자의 정보를 잠시동안 저장할 타입인 UserInfoTemporarySave 라는 클래스를 만들고 해당 클래스에 프로퍼티들과 채워졌는지 검사하는 함수, reset하는 함수를 구현하였습니다.

[구현한 부분]
1. UserInfoTemporarySave 타입 생성하여 step3에서 사용자가 입력하는 정보를 임시로 저장
2. keyboard에 accessoryview를 추가하여 done 버튼을 클릭시 키보드를 hide하도록 구현

jryoun1 committed:
feat: UserInfo Textfield placeholder 설정

jryoun1 committed:
feat: tapCance() 구현(cancel(), userInformationReset() 호출)

jryoun1 committed:
feat: imagePicker와 imagePickerController함수 구현

jryoun1 committed:
feat: imageView에 gestureRecognizer설정 및 tapImageView() 구현

jryoun1 committed:
refactor: informationTextView 색상 변경

jryoun1 committed:
refactor: originalImage가 아닌 user가 수정한 것을 반영한 editedImage로 수정

jryoun1 committed:
refactor: UIView의 contentMode scaleAspectFit -> scaleAspectFill로 수정

jryoun1 committed:
feat: UserInfoTemporarySave (유저 정보 임시 저장 타입)

jryoun1 committed:
feat: SignUpOptionViewController 세번째 VC

jryoun1 committed:
refactor: 3번째 회원가입화면 생성, 2번째 패스워드 secure, 다음 버튼 비활성화

jryoun1 committed:
feat: SignUpVC에서 tapGoNextButton() 구현

jryoun1 committed:
feat: SignUpVC에서 textfield, textview 입력 키보드에 done 버튼 추가

jryoun1 committed:
fix: SignUpVC NavigationBar숨기는 것을 viewWillAppear-> viewDidLoad에서 호출하도록 수정

매번 뷰가 전환될 때마다 굳이 바뀔 필요가 없이 처음에 로드될 때 선언해도 될 것 같아서 변경함

jryoun1 committed:
feat: isPasswordMatch() 패스워드 입력을 비교해주는 메소드 구현

jryoun1 committed:
feat: goNextButtonChanged() 다음버튼의 활성화상태 변경 메서드 구현

jryoun1 committed:
feat: 다음버튼 활성화 로직 구현

jryoun1 committed:
fix: 프로젝트 설정 Team, Bundle Indentifier 변경

AppleCEO commented on SignUpFlow/SignUpFlow/SignUpViewController.swift:
```diff
@@ -7,13 +7,141 @@
-class SignUpViewController: UIViewController {
-
+class SignUpViewController: UIViewController, UIImagePickerControllerDelegate,
+                            UINavigationControllerDelegate, UITextViewDelegate, UITextFieldDelegate {
```
프로토콜에 따라서 extension으로 나눠서 많이 사용합니다.
각 프로토콜에 대한 코드들을 모아놓아서 유지보수가 편합니다.
프로토콜마다 extension으로 나눠보세요.

AppleCEO commented on SignUpFlow/SignUpFlow/SignUpViewController.swift:
```diff
@@ -7,13 +7,141 @@
+    }
+    
+    func imagePickerController(_ picker: UIImagePickerController, didFinishPickingMediaWithInfo info: [UIImagePickerController.InfoKey : Any]) {
+        guard let editedImage: UIImage = info[UIImagePickerController.InfoKey.editedImage] as? UIImage else {
+            return
+        }
+        self.imageView.image = editedImage
+        self.dismiss(animated: true, completion: nil)
+        checkFilled.image = editedImage
+        goNextButtonEnableChange()
```
이 부분에 if let을 써보면 어떨까요? if let과 guard let을 각각 쓰기 좋을 때가 있습니다.
guard let만 고집하기보다 두가지를 병행해보다보면 어떤 때 if let이 좋다, 어떤 때가 guard let이 좋다 감이 올 것 같습니다.

jryoun1 replied:
@AppleCEO
사실 예전에도 guard와 if의 차이점에 대해서 리뷰를 받았던 적이 있어서 공부를 해봤습니다. 아직 확실하게 기준을 잡고 사용하고 있지는 않지만 Apple 문서에 의하면 if 보다 guard 를 사용하면 요구사항을 처리하는 코드와 에러를 처리하는 코드가 분리가 되어서 읽기가 더 수월하다고 나와있는 것을 보았습니다. 
이 경우에 대해서 다시 한번 생각해보니까 에러를 처리하는 코드를 작성한 것도 아니고, 사진의 정보를 가져오는 것이기 때문에 if let 을 사용하는 것이 더 괜찮을 것 같아요!!

AppleCEO commented on SignUpFlow/SignUpFlow/SignUpViewController.swift:
```diff
@@ -7,13 +7,141 @@
-    override func viewWillAppear(_ animated: Bool) {
-        navigationController?.isNavigationBarHidden = true
+    @IBAction private func tapCancelButton() {
+        cancel()
```
cancel과 같이 너무 단순한 메소드명은 똑같은 이름의 다른 메소드와 중복될 수 있습니다.
조금 더 상세한 메소드명이 좋을 것 같아요!

AppleCEO commented on SignUpFlow/SignUpFlow/SignUpViewController.swift:
```diff
@@ -7,13 +7,141 @@
+    func goNextButtonEnableChange() {
+        if checkFilled.isSignUpViewFilled() && isPasswordMatch() {
+            goNextButton.isEnabled = true
+        }
+        else {
+            goNextButton.isEnabled = false
+        }
```
} else { 이런 식으로 붙여서 사용하셨더라고요. 통일하시면 좋을 것 같습니다~

AppleCEO commented on SignUpFlow/SignUpFlow/SignUpViewController.swift:
```diff
@@ -7,13 +7,141 @@
+        self.view.endEditing(true)
+    }
+
+    private func cancel() {
+        self.dismiss(animated: true, completion: nil)
+    }
```
한줄짜리 코드를 수행하는 메소드를 만드셨네요!
이 코드를 그냥 다른 곳에 넣는 것과 메소드로 넣는 것의 차이는 무엇일까요?

jryoun1 replied:
@AppleCEO 
한 줄짜리 함수의 경우에는 만약 재 사용성이 많은 경우에는 따로 만드는 것이 재 사용성 측면에서 좋을 것 같지만 cancel() 함수는 사용하는 곳이 한 곳밖에 없어서 굳이 하나의 함수로 만들지 않고 tapCancelButton() 함수안에 넣어주어도 될 것 같아요!!

AppleCEO commented on SignUpFlow/SignUpFlow/UserInfoTemporarySave.swift:
```diff
@@ -0,0 +1,43 @@
+    var birthDate: Date?
+    
+    func isSignUpViewFilled() -> Bool {
+        if self.id?.isEmpty != nil && self.password?.isEmpty != nil && self.passwordCheck?.isEmpty != nil &&
+            self.image != nil && self.introduction?.isEmpty != nil {
+            return true
+        }
```
여기 조건문이 너무 복잡한데요.
어떤 식으로 좀 간결하게 할 수 있을까요?

AppleCEO reviewed (commented):
고생하셨습니다.
전체적으로 메소드명도 좋고 체계적이네요.
즐겁게 잘봤습니다.
몇가지 고민 점 남겨놨어요~

jryoun1 committed:
refactor: 프로토콜 마다 extension으로 나눠서 구현

jryoun1 committed:
refactor: imageView에 UITapGestureRecognizer 설정하는 코드를 함수로 구현

jryoun1 committed:
refactor: imagePickerController 에서 guard let -> if let으로 변경

이 경우에는 에러를 처리하는 코드를 작성한 것도 아니고, 사진의 정보를 가져오는 것이기 때문에 if let을 사용하는 것이 더 괜찮을 것 같아요

jryoun1 committed:
style: else문 { } 을 기존과 동일하게 수정

jryoun1 committed:
refactor: cancel() 함수를 제거하고 해당 코드를 tapCancelButton()함수로 추가

cancel() 함수는 한줄의 코드로 해당 뷰를 dismiss하는 코드인데 함수가 여러곳에서 호출된다면 함수로 유지해도 되지만 호출이 한번밖에 없으므로 함수를 제거하고 다른 함수에 해당 코드 추가

jryoun1 committed:
refactor: UserInfoTemporarySave의 isSignUpViewFilled(), isAllInfrFilled() 메서드 조건 검사 코드 수정

jryoun1 committed:
feat: yagom image 2x, 3x 추가

jryoun1 commented:
@AppleCEO 

[수정사항]
1. 프로토콜마다 extension으로 나눠서 구현
2. viewDidLoad()에서 tapGesture를 선언하고 imageview에 UITapRecognizer 설정하는 코드들을 빼서 setTapGesture() 메서드로 구현하고 viewDidLoad()에서는 setTapGesture()만 호출하도록 변경
3. imagePickerController에서 guard let -> if let으로 변경
4. cancel()함수 제거하고 tapCancelButton() 함수에 한줄 코드 추가
5. UserInfoTemporarySave 타입에서 isSignUpViewFilled(), isAllInfoFilled() 메서드에서 조건을 검사하는 부분 코드 수정

AppleCEO commented:
고생하셨습니다.
imagePicker에서 이미지를 선택했을 때 이미지의 비율이 유지가 안되고 퍼져보이는데요.
요구사항과 같이 이미지 비율을 맞춰서 뜨게 하려면 어떻게 할 수 있을까요?

AppleCEO commented on SignUpFlow/SignUpFlow/UserInfoTemporarySave.swift:
```diff
@@ -0,0 +1,51 @@
+    var phone: String?
+    var birthDate: Date?
+    
+    func isSignUpViewFilled() -> Bool {
+        guard let _ = self.id,
+              let _ = self.password,
+              let _ = self.passwordCheck,
+              let _ = self.image,
+              let _ = self.introduction else {
+            return false
+        }
+        return true
+    }
```
이 부분에서 값들이 있는지 확인하는군요!
체크하는 부분을 메소드로 따로 빼놓은 것은 아주 좋습니다.
guard let 구문 안에서 사용하지 않는 변수라면 와일드카드 패턴 말고 다른 방법으로 처리할 수 있을까요?

jryoun1 replied:
@AppleCEO
고민을 해봤는데 잘 모르겠어요ㅠㅠ 
어떠한 방법이 있을까요?? 🥲

AppleCEO commented on SignUpFlow/SignUpFlow/UserInfoTemporarySave.swift:
```diff
@@ -0,0 +1,51 @@
+    }
+    
+    func isAllInfoFilled() -> Bool {
+        if isSignUpViewFilled() {
+            guard let _ = self.birthDate,
+                  let _ = self.phone else {
+                return false
+            }
+            return true
+        } else {
+            return false
+        }
```
이 부분 흐름이 복잡하네요.
이렇게 리턴되는 곳이 많으면 더 이해하기 힘들어지는 것 같아요.
어떤 식으로 정리할 수 있을까요?

jryoun1 replied:
@AppleCEO
사실 isAllInfoFilled() 메서드는 step4에서 사용하려고 구현했습니다.
그리고 isSignUpViewFilled() 메서드는 두 번째 화면에서 정보가 채워지는 것을 확인하는 함수입니다. 
따라서 세 번째 화면으로 넘어갈 때 이미 isSignUpViewFilled를 체크하므로, 다시 체크할 필요가 없어서 삭제합니다!!

AppleCEO commented on SignUpFlow/SignUpFlow/SignUpViewController.swift:
```diff
@@ -7,13 +7,145 @@
+extension SignUpViewController: UITextViewDelegate {
+    func textViewDidChange(_ textView: UITextView) {
+        if introductionTextView.hasText {
+            guard let introduction = introductionTextView.text else { return }
```
고민해서 짜신 코드 같네요!
여기서는 else구문이 한줄에 나오고 있네요.
아래에서는 } else { 하고 한 줄씩 띄우고 있어요.
통일되는 것이 좋고 else 구문은 보통 아래 줄과 같이 한 줄 씩 띄어서 작성합니다.

AppleCEO commented on SignUpFlow/SignUpFlow/SignUpViewController.swift:
```diff
@@ -7,13 +7,145 @@
+        }
+        
+        if passwordCheckTextField.hasText {
+            guard let passwordCheck = passwordCheckTextField.text else { return }
```
옵셔널을 옵셔널 바인딩을 통해서 안전하게 꺼내는 것은 좋은 습관입니다 💯 
guard let 구문에 if let 을 적용해보면 어떨지 고민해보고 더 나은 선택지를 골라보세요!

jryoun1 replied:
@AppleCEO
이 부분에 관련해서 팀원인 라자냐도 비슷하지만 좀 다른 유형의 코드 리뷰를 받아서 같이 고민하다보니까 여기서는 If let을 써주는 것이 더 나을 것이라고 판단하게 되었어요!!
팀원인 라자냐의 경우에도 저와 비슷하게 guard let을 사용해서 text view, text field에서 nil이 들어오는 상황이 발생할 수 있을 것 같다고 생각했습니다. 하지만 Swift에서는 text view, text field에서는 nil을 대입해도 `""` 빈문자열이 대입된다는 것을 Apple 공식문서에서 확인해서 nil이 들어가는 경우가 없기 때문에 guard let을 사용하는 것보다는 If let을 사용하는 것이 나을 것 같다고 판단했습니다.

AppleCEO reviewed (commented):
고생하셨습니다!
imagePicker에서 이미지를 선택했을 때 이미지의 비율이 유지가 안되고 퍼져보이는데요.
요구사항과 같이 이미지 비율을 맞춰서 뜨게 하려면 어떻게 할 수 있을까요?
아래 내용들 확인해주세요 ^^

jryoun1 committed:
refactor: textfield, text view에서의 text를 옵셔널 바인딩 할 때 guard let-> if let으로 변경

Swift에서 text field, text view에 nil을 할당해도  빈문자열로 들어가기 때문에 guard let을 써서 예외처리를 하는 것보다는 if let을 사용하는 것이 더 적합하다고 판단

jryoun1 committed:
refactor: isAllInfoFilled() 메서드 isSignUpViewFilled()를 검사 조건에서 삭제하고 메서드 이름을 isAllInfoFilled->isOptionInfoFilled 로 수정

jryoun1 committed:
fix: imagePicker에서 비율이 AspectFill->AspectFit으로 변경

AppleCEO commented:
수정한 코드 있으면 푸시해주세요. 머지해드릴께요~

jryoun1 commented:
@AppleCEO
[수정사항]
1. AspectFill -> AspectFit으로 변경 (요구서에 비율을 유지한다고 해서 저는 Aspect를 사용한 Fit, Fill은 둘 다 비율은 유지되기 때문에 view의 크기에 맞춰주는 Fill을 사용했는데 Fill은 만약 빈 공간이 있을 경우에는 채워주면서 비율을 유지하기 때문에 확대되어 퍼져보이는 것 같아서 AspectFit으로 변경하였습니다)
2. guard let -> if let으로 변경
3. isAllInfoFilled() 메서드 명, 로직 변경

[고민했지만 해결하지 못한 부분]
와이드카드 패턴이 아닌 다른 방법으로 해결하는 방법을 고민해봤는데 잘 모르겠어요,,

AppleCEO commented:
제가 착각을 했는데요. 
프로필 사진은 화면이 꽉 차게 보여야하기 때문에 AspectFill 로 다시 바꿔서 다음 단계 진행하시면 될 것 같습니다.
착오 일으켜서 미안해요.
